ci(build-test): add condition for md files

DariuszPorowski opened the pull request:
# Description

This pull request updates the GitHub Actions workflows to optimize CI/CD runs by skipping unnecessary jobs when only Markdown documentation files are changed in selected directories. It introduces a new `changes` job to detect Markdown-only changes and uses its output to conditionally run build, test, and deployment jobs. Additionally, it updates Node.js versions and makes minor improvements to workflow configuration.

## Type of change

<!--

Please select **one** of the following options that describes your change and delete the others. Clearly identifying the type of change you are making will help us review your PR faster, and is used in authoring release notes.

If you are making a bug fix or functionality change to Radius and do not have an associated issue link please create one now. 

-->

- This pull request is a minor refactor, code cleanup, test improvement, or other maintenance task and doesn't change the functionality of Radius (issue link optional).

## Contributor checklist
Please verify that the PR meets the following requirements, where applicable:

<!--
This checklist uses "TaskRadio" comments to make certain options mutually exclusive.
See: https://github.com/mheap/require-checklist-action?tab=readme-ov-file#radio-groups
For details on how this works and why it's required.
-->

- An overview of proposed schema changes is included in a linked GitHub issue.
    - [ ] Yes <!-- TaskRadio schema -->
    - [x] Not applicable <!-- TaskRadio schema -->
- A design document PR is created in the [design-notes repository](https://github.com/radius-project/design-notes/), if new APIs are being introduced.
    - [ ] Yes <!-- TaskRadio design-pr -->
    - [x] Not applicable <!-- TaskRadio design-pr -->
- The design document has been reviewed and approved by Radius maintainers/approvers.
    - [ ] Yes <!-- TaskRadio design-review -->
    - [x] Not applicable <!-- TaskRadio design-review -->
- A PR for the [samples repository](https://github.com/radius-project/samples) is created, if existing samples are affected by the changes in this PR.
    - [ ] Yes <!-- TaskRadio samples-pr -->
    - [X] Not applicable <!-- TaskRadio samples-pr -->
- A PR for the [documentation repository](https://github.com/radius-project/docs) is created, if the changes in this PR affect the documentation or any user facing updates are made.
    - [ ] Yes <!-- TaskRadio docs-pr -->
    - [X] Not applicable <!-- TaskRadio docs-pr -->
- A PR for the [recipes repository](https://github.com/radius-project/recipes) is created, if existing recipes are affected by the changes in this PR.
    - [ ] Yes <!-- TaskRadio recipes-pr -->
    - [X] Not applicable <!-- TaskRadio recipes-pr -->

DariuszPorowski committed:
ci(build-test): add conditions for md files

Signed-off-by: Dariusz Porowski <3431813+DariuszPorowski@users.noreply.github.com>

DariuszPorowski commented on .github/workflows/functional-test-noncloud.yaml:
```diff
@@ -87,9 +87,37 @@ env:
+              - ".github/**/*.md"
+              - "docs/**/*.md"
+              - "docs/*.md"
+
```
@copilot `md_only_changed` exists, it's dynamic output based on the filter key

brooke-hamilton reviewed (changes requested):
🥇 This PR is going to be a huge help to reducing the wait times on documentation updates 🥇 

Should we also add the conditional check to the `functional-tests-approval.yaml` workflow?

DariuszPorowski committed:
ci(changes): refactor job definitions to use reusable workflow

Signed-off-by: Dariusz Porowski <3431813+DariuszPorowski@users.noreply.github.com>

nithyatsu commented on .github/workflows/__changes.yml:
```diff
@@ -0,0 +1,48 @@
+          test/**/*.md
+          test/*.md
+          *.md
+    outputs:
```
We would still generate/build/test when a README.md such as the one mentioned in `/radius/build/generate.mk` changes correct? One example line below:

```
generate-rad-corerp-client: generate-node-installed generate-autorest-installed generate-tsp-installed generate-openapi-spec ## Generates the corerp client SDK (Autorest).
	@echo "$(ARROW) Generating 'pkg/corerp/api/v20231001preview'"
	autorest pkg/corerp/api/README.md --tag=core-2023-10-01-preview && rm pkg/corerp/api/v20231001preview/go.mod && go fmt ./pkg/corerp/api/v20231001preview/...
	@echo "$(ARROW) Done."
```

DariuszPorowski committed:
ci(build): remove unused `az login` (leftover after #8565) (#10849)

Flexible federated identity credentials setup is not required. It was
unused code, and migration is not required. Leftover after #8565

bonus:
- style: basic lint

<!--

Please select **one** of the following options that describes your
change and delete the others. Clearly identifying the type of change you
are making will help us review your PR faster, and is used in authoring
release notes.

If you are making a bug fix or functionality change to Radius and do not
have an associated issue link please create one now.

-->

- This pull request adds or changes features of Radius and has an
approved issue (issue link required).

<!--

Please update the following to link the associated issue. This is
required for some kinds of changes (see above).

-->

Fixes: #10796

Please verify that the PR meets the following requirements, where
applicable:

<!--
This checklist uses "TaskRadio" comments to make certain options
mutually exclusive.
See:
https://github.com/mheap/require-checklist-action?tab=readme-ov-file#radio-groups
For details on how this works and why it's required.
-->

- An overview of proposed schema changes is included in a linked GitHub
issue.
    - [ ] Yes <!-- TaskRadio schema -->
    - [x] Not applicable <!-- TaskRadio schema -->
- A design document PR is created in the [design-notes
repository](https://github.com/radius-project/design-notes/), if new
APIs are being introduced.
    - [ ] Yes <!-- TaskRadio design-pr -->
    - [x] Not applicable <!-- TaskRadio design-pr -->
- The design document has been reviewed and approved by Radius
maintainers/approvers.
    - [ ] Yes <!-- TaskRadio design-review -->
    - [x] Not applicable <!-- TaskRadio design-review -->
- A PR for the [samples
repository](https://github.com/radius-project/samples) is created, if
existing samples are affected by the changes in this PR.
    - [ ] Yes <!-- TaskRadio samples-pr -->
    - [x] Not applicable <!-- TaskRadio samples-pr -->
- A PR for the [documentation
repository](https://github.com/radius-project/docs) is created, if the
changes in this PR affect the documentation or any user facing updates
are made.
    - [ ] Yes <!-- TaskRadio docs-pr -->
    - [x] Not applicable <!-- TaskRadio docs-pr -->
- A PR for the [recipes
repository](https://github.com/radius-project/recipes) is created, if
existing recipes are affected by the changes in this PR.
    - [ ] Yes <!-- TaskRadio recipes-pr -->
    - [X] Not applicable <!-- TaskRadio recipes-pr -->

---------

Signed-off-by: Dariusz Porowski <3431813+DariuszPorowski@users.noreply.github.com>

DariuszPorowski committed:
ci(deps): update actions

Signed-off-by: Dariusz Porowski <3431813+DariuszPorowski@users.noreply.github.com>
